hibernate-core/src/main/java/org/hibernate/sql/ast/spi/AbstractSqlAstTranslator.java

```diff
@@ -1092,6 +1095,59 @@ protected Predicate determineWhereClauseRestrictionWithJoinEmulation(AbstractUpd
 		querySpec.getSelectClause().addSqlSelection(
 				new SqlSelectionImpl( new QueryLiteral<>( 1, getIntegerType() ) )
 		);
+
+		final MutableObject<StandardVirtualTableGroup> virtualTableGroup;
+		final List<TableGroupJoin> collectedNonInnerJoins;
+		if ( supportsJoinInMutationStatementSubquery() ) {
+			virtualTableGroup = new MutableObject<>();
+			collectedNonInnerJoins = new ArrayList<>();
+			emulateWhereClauseRestrictionJoins( statement, querySpec, tableGroupJoin -> {
+				if ( tableGroupJoin.getJoinType() == SqlAstJoinType.INNER ) {
+					final TableGroup joinedGroup = tableGroupJoin.getJoinedGroup();
+					if ( !virtualTableGroup.isSet() ) {
+						virtualTableGroup.set( new StandardVirtualTableGroup(
+								joinedGroup.getNavigablePath(),
+								joinedGroup.getModelPart(),
+								joinedGroup,
+								joinedGroup.isFetched()
+						) );
+					}
+					virtualTableGroup.get().addTableGroupJoin( tableGroupJoin );
+					querySpec.applyPredicate( tableGroupJoin.getPredicate() );
+				}
+				else {
+					collectedNonInnerJoins.add( tableGroupJoin );
+				}
+			} );
+		}
+		else {
+			virtualTableGroup = null;
+			collectedNonInnerJoins = null;
+			emulateWhereClauseRestrictionJoins( statement, querySpec, tableGroupJoin -> {
+				if ( tableGroupJoin.getJoinType() == SqlAstJoinType.INNER ) {
+					querySpec.getFromClause().addRoot( tableGroupJoin.getJoinedGroup() );
+					querySpec.applyPredicate( tableGroupJoin.getPredicate() );
+				}
+			} );
+		}
+
+		if ( virtualTableGroup != null && virtualTableGroup.isSet() ) {
+			final TableGroup firstInnerJoinedGroup = virtualTableGroup.get().getUnderlyingTableGroup();
+			collectedNonInnerJoins.forEach( firstInnerJoinedGroup::addTableGroupJoin );
+			querySpec.getFromClause().addRoot( virtualTableGroup.get() );
+		}
+		else if ( querySpec.getFromClause().getRoots().isEmpty() ) {
+			return statement.getRestriction();
+		}
```

Comment:
```suggestion
		if ( querySpec.getFromClause().getRoots().isEmpty() ) {
			return statement.getRestriction();
		}
		else if ( collectedNonInnerJoins != null ) {
			final TableGroup firstInnerJoinedGroup = querySpec.getFromClause().getRoots().get( 0 );
			for ( int i = 0; i < collectedNonInnerJoins.size(); i++ ) {
				firstInnerJoinedGroup.addTableGroupJoin( collectedNonInnerJoins.get( i ) );
			}
		}
```

Reply:
This won't work because we would be applying `collectedNonInnerJoins` on the `StandardVirtualTableGroup` we created earlier. This would cause them to get rendered as roots, instead of actual e.g. `left join`s. That's why I added the joins on the underlying table group in the first place.

Reply:
Altering existing parts of the SQL AST (e.g. `TableGroup`) is unfortunately also not going to work. The AST is cached, so any changes you do are racy from a concurrency perspective and what you're doing here is especially problematic because the transformation is not even idempotent.
Maybe you'll simply have to create a copy of the first table group as `StandardTableGroup` instead of a  `StandardVirtualTableGroup` then?